### PR TITLE
Added a pre run atom for checking validity of node

### PR DESCRIPTION
**Pull request by shtripat**
Before starting the import/create cluster we should do
a pre-check for the validity and availability of node
entity and its details.

tendrl-bug-id: Tendrl/commons#550
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by r0h4n** [comment]
Please file a issue on commons for this describing details

**Comment by shtripat** [comment]
>> Please file a issue on commons for this describing details, and attach it to commit msg via
>> "tendrl-bug-id: Tendrl/commons#issue_id"

Done

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            try:
+                NS._int.client.read("nodes/%s/Memory" % node_id)
+            except etcd.EtcdKeyNotFound:
+                raise AtomExecutionFailedError(
+                    "Node doesnt have Memory details populated"
+                )
+
+            # Check if node has the block devices populated
+            try:
+                NS._int.client.read("nodes/%s/BlockDevices" % node_id)
```
Comment (r0h4n) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: You'll have to check /Disks too.

Also, youll need to add checks if these /Disks or any other directories exist and are **not empty**  (meaning result.leaves is not None)

File: tendrl/commons/flows/create_cluster/__init__.py [hunk 2]
```diff
@@ -85,7 +85,9 @@ def run(self):
                     NS.node_context.save()
                 break
 
-                                               
+        # Execute super run() to execute pre-runs
+        super(CreateCluster, self).run()
```
Comment (shtripat) [inline tendrl/commons/flows/create_cluster/__init__.py]: Ok. Will move it in the start of the run() function.

File: tendrl/commons/flows/import_cluster/__init__.py [hunk 3]
```diff
@@ -104,6 +103,9 @@ def run(self):
                         self.parameters['gdeploy_provisioned'] = True
                         break
 
+        # Invoke the super run() for pre-runs execution
+        super(ImportCluster, self).run()
```
Comment (shtripat) [inline tendrl/commons/flows/import_cluster/__init__.py]: Done

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import etcd
+
+from tendrl.commons import objects
+from tendrl.commons.objects import AtomExecutionFailedError
+from tendrl.commons.objects.node import Node
+
+
+class ValidNodeDetails(objects.BaseAtom):
+    obj = Node
```
Comment (shtripat) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: Done. Will remove this.

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py (hunk 1 above)
Comment (shtripat) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: Ok. will add the checks for /Disks as well.

File: tendrl/commons/flows/create_cluster/__init__.py [hunk 5]
```diff
@@ -42,6 +42,9 @@ def run(self):
                 "Space char not allowed in cluster name"
             )
 
+        # Execute super run() to execute pre-runs
```
Comment (r0h4n) [inline tendrl/commons/flows/create_cluster/__init__.py]: add to comment that "atoms and post_runs will also be executed due to this super call"

also add a "TODO (team) break down run() into run_pre(), run_atom(), run_post() where we run the pre_runs, atoms, post_runs respectively so run() simply calls run_pre(), run_atom(), run_post()

File: tendrl/commons/flows/import_cluster/__init__.py [hunk 6]
```diff
@@ -15,12 +15,14 @@
...
 class ImportCluster(flows.BaseFlow):
     def run(self):
-
         integration_id = self.parameters['TendrlContext.integration_id']
         if integration_id is None:
             raise FlowExecutionFailedError("TendrlContext.integration_id cannot be empty")
         sds_name = self.parameters['DetectedCluster.sds_pkg_name']
 
+        # Invoke the super run() for pre-runs execution
+        super(ImportCluster, self).run()
```
Comment (r0h4n) [inline tendrl/commons/flows/import_cluster/__init__.py]: same as comment for create cluster

File: tendrl/commons/objects/definition/master.yaml [hunk 7]
```diff
@@ -37,6 +37,8 @@ namespace.tendrl:
           - TendrlContext.integration_id
           - Cluster.node_configuration
           - Cluster.conf_overrides
+      pre_run:
+        - tendrl.objects.Node.atoms.ValidNodeDetails
```
Comment (r0h4n) [inline tendrl/commons/objects/definition/master.yaml]: If suggest we change the atom name from "ValidNodeDetails" to "IsNodeTendrlManaged" 

which gives us an atom to do all checks we can if a certain node is Tendrl managed

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+import etcd
+
+from tendrl.commons import objects
+from tendrl.commons.objects import AtomExecutionFailedError
+from tendrl.commons.objects.node import Node
```
Comment (r0h4n) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: remove this import, unused

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            except etcd.EtcdKeyNotFound:
+                raise AtomExecutionFailedError(
+                    "Node doesnt have block device details populated"
+                )
+
+            # Check if node has the disks data populated
+            try:
+                disks = NS._int.client.read("nodes/%s/Disks" % node_id)
+                # If disks have no child nodes, error out
+                if len(disks._children) == 0:
```
Comment (r0h4n) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: @shtripat its a bad practice to use private vars of third party deps (_children), moreover i did mention in my comment to use .leaves 

Also, you need to check if all items like Os, memory etc are not empty

File: tendrl/commons/flows/create_cluster/__init__.py (hunk 5 above)
Comment (shtripat) [inline tendrl/commons/flows/create_cluster/__init__.py]: sure.

File: tendrl/commons/objects/definition/master.yaml (hunk 7 above)
Comment (shtripat) [inline tendrl/commons/objects/definition/master.yaml]: done

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py (hunk 8 above)
Comment (shtripat) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: done

File: tendrl/commons/objects/node/atoms/valid_node_details/__init__.py (hunk 9 above)
Comment (shtripat) [inline tendrl/commons/objects/node/atoms/valid_node_details/__init__.py]: done

**Comment by r0h4n** [comment]
Verified?

**Comment by shtripat** [comment]
@r0h4n I have verified this with import cluster flow.

**Pull request merged**